package/src/ssr/NextSSRApolloClient.tsx

```diff
@@ -0,0 +1,169 @@
+  }
+
+  private registerWindowHook() {
+    if (typeof window !== "undefined") {
+      if (Array.isArray(window[ApolloBackgroundQueryTransport] || [])) {
+        registerLateInitializingQueue(
+          ApolloBackgroundQueryTransport,
+          (options) => {
+            const { query, varJson, cacheKey } =
+              this.identifyUniqueQuery(options);
+
+            const byVariables =
+              this["queryManager"].inFlightLinkObservables.get(query) ||
+              new Map();
+
+            this["queryManager"].inFlightLinkObservables.set(
+              query,
+              byVariables
+            );
+
+            if (!byVariables.has(varJson)) {
+              const promise = new Promise<FetchResult>((resolve, reject) => {
+                this.resolveFakeQueries.set(cacheKey, [resolve, reject]);
+              });
+
+              byVariables.set(
+                varJson,
+                new Observable<FetchResult>((observer) => {
+                  promise
+                    .then((result) => {
+                      observer.next(result);
+                      observer.complete();
+                    })
+                    .catch((err) => {
+                      observer.error(err);
+                    });
+                })
+              );
+
+              const cleanupCancelFn = () =>
+                this["queryManager"].fetchCancelFns.delete(cacheKey);
+
+              const [_, reject] = this.resolveFakeQueries.get(cacheKey) ?? [];
+
+              this["queryManager"].fetchCancelFns.set(
+                cacheKey,
+                (reason: unknown) => {
+                  cleanupCancelFn();
+                  if (reject) {
+                    this.resolveFakeQueries.delete(cacheKey);
+                    reject(reason);
+                  }
+                }
+              );
+            }
+          }
+        );
+      }
+
+      if (Array.isArray(window[ApolloResultCache] || [])) {
+        registerLateInitializingQueue(ApolloResultCache, (data) => {
+          const { cacheKey } = this.identifyUniqueQuery(data);
+          const [resolve] = this.resolveFakeQueries.get(cacheKey) ?? [];
+
+          if (resolve) {
+            resolve({
+              data: data.result,
+            });
+            this.resolveFakeQueries.delete(cacheKey);
+          }
+        });
+      }
+    }
+  }
+
+  watchQuery<
+    T = any,
+    TVariables extends OperationVariables = OperationVariables
+  >(options: WatchQueryOptions<TVariables, T>) {
+    if (typeof window == "undefined") {
+      this.rehydrationContext.incomingBackgroundQueries.push(options);
+    }
+    const result = super.watchQuery(options);
+    return result;
+  }
+
+  setRehydrationContext(rehydrationContext: RehydrationContextValue) {
+    if (this.rehydrationContext.uninitialized) {
+      rehydrationContext.incomingBackgroundQueries.push(
+        ...this.rehydrationContext.incomingBackgroundQueries
+      );
+    }
```

Comment:
Maybe here we could even move all data from the old `incombineBackgroundQueries` to the new ones?

Something like

```suggestion
    if (rehydrationContext.incomingBackgroundQueries !== this.rehydrationContext.incomingBackgroundQueries)
      rehydrationContext.incomingBackgroundQueries.push(
        ...this.rehydrationContext.incomingBackgroundQueries.splice(0, Infinity)
      )
```

Then we could skip all the `uninitialized` handling altogether.

Reply:
Fixed in [c49d481](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/c49d481ff5c9bdd474c4b61ba65b610162ffe067)